### ISSUE TITLE
Parametrized parser for speed

### ISSUE THREAD

**Issue by asinghvi17** [evetion/WellKnownGeometry.jl#30]
It would be great to have a parameterized parser object for speed.  Things we could specialize on are:
- Endianness (little->big endian is just `bswap` on the result Float64 or Int)
- Geometry type: when parsing SQL query results, we often know that the column contains only one kind of geometry.  In this case, it would be nice to be able to call a parser which only parses, e.g., points, or polygons, etc.  This parameterized parser could simply error if the geometry was found to be of an incorrect type.  This also guarantees return type and would probably make fast inner loops possible when parsing columns in [GeoParquet](https://github.com/JuliaGeo/GeoParquet.jl/blob/655cc5b2cfb54f5de388071740c57950c392a1a0/src/io.jl#L52) and [GeoPackage](https://github.com/JuliaGeo/GeoPackage.jl).
	- The generic parser would be type unstable and do all necessary type checking, similar to the current approach.

**Comment by rafaqz** [comment]
We also know when it contains a small union, and a function with a conditional like this should allow both optimisations:

```julia
function wrap_wkb(::Type{Trait}, ::Val{HasZ}, Val{HasM}, wkb) where {Trait, HasZ, HasM}
    if PolygonTrait <: T && _get_wkb_typenum(wkb) == POLYGON_CODE
        return WKBPolygon{HasZ,HasM}(wkb)
    elseif MultiPolygonTrait <: T &&  _get_wkb_typenum(wkb) == MULTIPOLYGON_CODE
        return WKBMultiPolygon{HasZ,HasM}(wkb)
    elseif MultiPolygonTrait <: T &&  _get_wkb_typenum(wkb) == MULTIPOINT_CODE
        return WKBMultiPoint{HasZ,HasM}(wkb)
    ...
    end
end
```

The compiler should be able to delete the code for branches not in the type union.

The idea is this would be used on a a vector/column of geometries after a function barrier where the types were worked out for the whole column. In GeoPacakage this can be done with `union` on the geometry type column as Integers, the a conversion to types. So one point of instability.

Nested geoms can use views into the wkb, maybe with some type of runtime information to deal with the fact some data is missing from child objects (like endianness) that is present if they're at the top layer.

This should be able to mean all the cost is at the top layer and `wrap_wkb` is essentially free wrappers over the wkb data. The cost will increase as the type union gets larger and we lose first stability and then small union optimisations.

**Comment by rafaqz** [comment]
Oh cool. Absolutely. Is it already stored like that?

**Comment by evetion** [comment]
You need to parse some stuff for the dimensions and the number of coordinates, but sure. I use reinterpret all over (in serializing mostly), so this would work fine.

**Comment by rafaqz** [comment]
I think we want to do that outside the function barrier and then maybe check at each geometry that it matches, and either error or `reinterpret`

**Comment by asinghvi17** [comment]
At least for values of "maybe has M" or "maybe has Z" it would have to be dynamic, but when it's defined that the geometry definitely has M or Z, that approach should be fine AFAICT.

We would have to allow the user to choose whether to use a strict or dynamic parser in this case, since it's possible that people made flawed datasets.

**Comment by asinghvi17** [comment]
`ls_as_wkb` is a WKB of 300,000 points
```julia
# Parsing approach 2 - reduce to reinterpret
@benchmark collect(reinterpret(Tuple{Float64, Float64}, view($ls_as_wkb, (1+4+4+1):length($ls_as_wkb))))

# BenchmarkTools.Trial: 10000 samples with 334 evaluations.
#  Range (min … max):  258.111 ns …  3.375 μs  ┊ GC (min … max): 0.00% … 89.28%
#  Time  (median):     274.826 ns              ┊ GC (median):    0.00%
#  Time  (mean ± σ):   280.685 ns ± 97.172 ns  ┊ GC (mean ± σ):  1.11% ±  2.94%

#            ▁▄█▇▅▂  ▁ ▂▁
#   ▁▁▂▃▃▃▄▃▅███████▇████▆▆▄▄▄▄▃▃▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃
#   258 ns          Histogram: frequency by time          326 ns <

#  Memory estimate: 160 bytes, allocs estimate: 4.
```

which is not bad!  `GFT.WellKnownBinary(GFT.Geom(), ls_as_wkb) |> GI.getpoint |> collect` is still running as I write this.

**Comment by evetion** [comment]
> `ls_as_wkb` is a WKB of 300,000 points

Nice! Can we create a repo, or attach a few of these binary test/benchmark files to a release, so we can add this to CI, and test ourselves as well?

**Comment by asinghvi17** [comment]
In this case it's as simple as:
```julia
import GeoFormatTypes as GFT, WellKnownGeometry as WKG, GeoInterface as GI

tups = tuple.(rand(300_000), rand(300_000))
geoms_as_points = GFT.val.(WKG.getwkb.(GI.Point.(tups)))
geoms_as_linestring = GFT.val(WKG.getwkb(GI.LineString(tups)))
```
and try the different parsing approaches from there.

```julia
# parse geoms as points, each WKB individually
@benchmark begin
    map($(geoms_as_points)) do geom
        only(reinterpret(Tuple{Float64, Float64}, view(geom, (5+1):length(geom)))) # this has to be parsed differently in EWKB and GeoPkg wkb
    end
end # median time 3ms v/s regular parsing 300ms

# parse the whole linestring
@benchmark collect(reinterpret(Tuple{Float64, Float64}, view($ls_as_wkb, (1+4+4+1):length($ls_as_wkb))))
# median time 290ns vs unknown (still running) 
# we could even drop collect here
```

**Comment by asinghvi17** [comment]
Looks like `GI.getpoint` hangs because it's calling `ncoord` every point...
 [2] getgeom(T::GeoInterface.LineStringTrait, geom::GeoFormatTypes.WellKnownBinary{GeoFormatTypes.Geom, Vector{UInt8}}, i::Int64)
   @ WellKnownGeometry ~/.julia/packages/WellKnownGeometry/t1n94/src/wkb.jl:190

We could optimize that pretty easily no?

**Comment by evetion** [comment]
If you know it's a line string, or know the dimension, yes, absolutely.

**Comment by asinghvi17** [comment]
Huh, no this just worked for me...let me try it in a fresh session!

**Comment by asinghvi17** [comment]
No, I'm on latest released GeoInterface and WKG v0.2.2 (one patch version behind)

**Comment by evetion** [comment]
Ouch, my bad then it seems.

**Comment by asinghvi17** [comment]
Monkeypatching
```julia
GI.ncoord(::GI.Wrappers.WrapperGeometry{false, false}) = 2
GI.ncoord(::GI.Wrappers.WrapperGeometry{false, true}) = 3
GI.ncoord(::GI.Wrappers.WrapperGeometry{true, false}) = 3
GI.ncoord(::GI.Wrappers.WrapperGeometry{true, true}) = 4
```
fixes it for now, but this probably needs a better check.  Maybe `GI.is3d` and `GI.ismeasured`?

**Comment by evetion** [comment]
> `ls_as_wkb` is a WKB of 300,000 points
> 
> ```julia
> # Parsing approach 2 - reduce to reinterpret
> @benchmark collect(reinterpret(Tuple{Float64, Float64}, view($ls_as_wkb, (1+4+4+1):length($ls_as_wkb))))
> 
> # BenchmarkTools.Trial: 10000 samples with 334 evaluations.
> #  Range (min … max):  258.111 ns …  3.375 μs  ┊ GC (min … max): 0.00% … 89.28%
> #  Time  (median):     274.826 ns              ┊ GC (median):    0.00%
> #  Time  (mean ± σ):   280.685 ns ± 97.172 ns  ┊ GC (mean ± σ):  1.11% ±  2.94%
> 
> #            ▁▄█▇▅▂  ▁ ▂▁
> #   ▁▁▂▃▃▃▄▃▅███████▇████▆▆▄▄▄▄▃▃▂▂▂▂▂▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁▁ ▃
> #   258 ns          Histogram: frequency by time          326 ns <
> 
> #  Memory estimate: 160 bytes, allocs estimate: 4.
> ```
> 
> which is not bad! `GFT.WellKnownBinary(GFT.Geom(), ls_as_wkb) |> GI.getpoint |> collect` is still running as I write this.

For what it's worth, `getpoint`, gets you a point, as in creating a new WKB Point. You probably want `coordinates` instead, if it is written roughly like this:

```
function GI.coordinates(
    T::GI.LineStringTrait,
    geom::WKBtype,
)
    size = headersize + numsize
    ncoord = GI.ncoord(geom)
    offset = typesize(wkbsubtype(T), geom[size+1:end], ncoord)
    reinterpret(NTuple{ncoord,Float64}, @view geom.val[size+1:size+offset*GI.ngeom(geom)])
end
```

Which while still a bit unstable with ncoord, is already doable with 8.958 μs (15 allocations: 469.34 KiB).

**Comment by rafaqz** [comment]
@asinghvi17 do you  want to PR your patch to GI? I must have forgotten to add it.

**Comment by evetion** [comment]
Yeah, it probably needs to be a tuple. And if the type has noop ncoord it should be pretty fast, even without refactoring it everywhere.
